# Data Analytics Program¶
## Python Fundamentals for Data Analysts
## 4.6: Combining & Exporting data
## Alphy Jose

# Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# Step 3:import the orders_products_combined dataframe from the pickle file you just saved.

In [2]:
# Creating string of project folder path
path = r'C:\Users\augus\Documents\01-2026 Instacart Basket Analysis'

In [3]:
#importing orders_products_combined dataframe from the pickle file
df_ords_prds_comb = pd.read_pickle(os.path.join(path, '02 Data' , 'Prepared Data' , 'orders_products_combined.pkl'))

In [6]:
df_ords_prds_comb.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     NaN         196                  1          0   both  
1                     NaN       14084                  2          0   both  
2                     NaN       12427                  3          0   both  
3                     NaN       26088                  4          0   both  
4                     NaN       26405                  5          0   both

# Step 4: Check the shape of the imported dataframe

In [5]:
df_ords_prds_comb.shape

(32434489, 10)

#### The shape(rows & columns) is same

# Step 5:Determine a suitable way to combine the orders_products_combined dataframe with your products data set.

In [4]:
#import the product dataframe from prepared data file.
df_prds= pd.read_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'),index_col=0)

In [7]:
df_prds.head()

product_name  aisle_id  \
product_id                                                                
1                                  Chocolate Sandwich Cookies        61   
2                                            All-Seasons Salt       104   
3                        Robust Golden Unsweetened Oolong Tea        94   
4           Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
5                                   Green Chile Anytime Sauce         5   

            department_id  prices  
product_id                         
1                      19     5.8  
2                      13     9.3  
3                       7     4.5  
4                       1    10.5  
5                      13     4.3

In [10]:
df_prds.shape

(49672, 4)

In [9]:
df_prds.describe()

aisle_id  department_id        prices
count  49672.000000   49672.000000  49672.000000
mean      67.762442      11.728942      7.680414
std       38.315784       5.850779      4.199298
min        1.000000       1.000000      1.000000
25%       35.000000       7.000000      4.100000
50%       69.000000      13.000000      7.100000
75%      100.000000      17.000000     11.100000
max      134.000000      21.000000     25.000000

In [12]:
# Test merge without overwriting
pd.merge(df_ords_prds_comb,df_prds, on = 'product_id',indicator = True)

ValueError: Cannot use name of an existing column for indicator column

In [6]:
# dropping _merge column from df_ords_prds_comb dataframe because merging will create a new indicator column with same name and its is showing and error
df_ords_prds_comb = df_ords_prds_comb.drop(['_merge'], axis=1)

In [16]:
df_ords_prds_comb.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  
0                     NaN         196                  1          0  
1                     NaN       14084                  2          0  
2                     NaN       12427                  3          0  
3                     NaN       26088                  4          0  
4                     NaN       26405                  5          0

In [14]:
# Test merge without overwriting
pd.merge(df_ords_prds_comb,df_prds, on = 'product_id',indicator = True)

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2539329        1             1                   2   
2          2539329        1             1                   2   
3          2539329        1             1                   2   
4          2539329        1             1                   2   
...            ...      ...           ...                 ...   
32404854   2977660   206209            13                   1   
32404855   2977660   206209            13                   1   
32404856   2977660   206209            13                   1   
32404857   2977660   206209            13                   1   
32404858   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         8                     NaN       14084   
2                         8                     NaN       12427   
3                         8                     NaN       26088   
4                         8                     NaN       26405   
...                     ...                     ...         ...   
32404854                 12                     7.0       14197   
32404855                 12                     7.0       38730   
32404856                 12                     7.0       31477   
32404857                 12                     7.0        6567   
32404858                 12                     7.0       22920   

          add_to_cart_order  reordered  \
0                         1          0   
1                         2          0   
2                         3          0   
3                         4          0   
4                         5          0   
...                     ...        ...   
32404854                  5          1   
32404855                  6          0   
32404856                  7          0   
32404857                  8          0   
32404858                  9          0   

                                     product_name  aisle_id  department_id  \
0                                            Soda        77              7   
1         Organic Unsweetened Vanilla Almond Milk        91             16   
2                             Original Beef Jerky        23             19   
3                      Aged White Cheddar Popcorn        23             19   
4                XL Pick-A-Size Paper Towel Rolls        54             17   
...                                           ...       ...            ...   
32404854                             Tomato Paste         9              9   
32404855          Brownie Crunch High Protein Bar         3             19   
32404856    High Protein Bar Chunky Peanut Butter         3             19   
32404857      Chocolate Peanut Butter Protein Bar         3             19   
32404858      Roasted & Salted Shelled Pistachios       117             19   

          prices _merge  
0            9.0   both  
1           12.5   both  
2            4.4   both  
3            4.7   both  
4            1.0   both  
...          ...    ...  
32404854     5.6   both  
32404855     5.9   both  
32404856     4.2   both  
32404857     4.9   both  
32404858     1.7   both  

[32404859 rows x 14 columns]

In [24]:
# merging both dataframes to create df_ords_prods_merge. I am not explicitly providing the 'how' argument because the 
# default join is inner join.
df_ords_prods_merge= df_ords_prds_comb.merge(df_prds, on = 'product_id', indicator = True)

#### The order of DataFrames doesnt makes a difference in the resulting dataframe when using the merge function in pandas. 
#### The core data (which rows are kept) should be consistent regardless of order.
#### The difference is only in terms of the column order in the output DataFrame and the handling of overlapping columns not used as keys. 
#### I am gvining df_ords_prds_comb dataframe first , so that the resulting dataframe will have order details first
#### And there are no columns with same name other than the key in both dataframes.

In [29]:
df_ords_prods_merge.shape

(32404859, 14)

#### The number of rows is less than df_ords_prds_comb dataset. So each of those rows doesnt have information  in both input data sets.

In [26]:
df_ords_prods_merge.head(1)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both

# Step 6:Confirm the results of the merge using the merge flag.

In [19]:
#checking the result of merge using merge flag
df_ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

#### In the “_merge” column, we can see that there are only entries that have a value of “both"
#### I am testing actual merge rate with outer join 

In [7]:
# merging both dataframes with outer join
df_ords_prods_merge_test= df_ords_prds_comb.merge(df_prds, on = 'product_id',how = 'outer',indicator = True)

In [8]:
df_ords_prods_merge_test['_merge'].value_counts()

_merge
both          32404859
left_only        30200
right_only          11
Name: count, dtype: int64

In [10]:
unique_orders_ids = df_ords_prods_merge_test[df_ords_prods_merge_test['_merge'] == 'left_only']['order_id'].nunique()
print(unique_orders_ids)

30066


#### There are 30066 orders without product data and 11 products that were never orderd
#### For this project I am proceeding with same inner join merge since we are working with datasets with full merge rate

# Step 7:Export the newly created dataframe as ords_prods_merge in a suitable format

In [28]:
# Export dataframe  in pkl format
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge.pkl'))